jira_sync/sync_mgr.py

```diff
@@ -0,0 +1,303 @@
+        return full_url
+
+    @classmethod
+    def match_jira_forge_issues(
+        cls, jira_issues: Collection[JiraIssue], forge_issues: Collection[Issue]
+    ) -> tuple[set[MatchedIssue], set[JiraIssue], set[Issue]]:
+        """Match JIRA issues with those from forges.
+
+        :param jira_issues: The JIRA issues to be matched up
+        :param forge_issues: The forge issues to be matched up
+
+        :returns: Matched pairs of JIRA and forge issues, unmatched jira
+            issues and unmatched forge issues
+        """
+        log.info("Matching JIRA and forge issues…")
+        matched_issues = set()
+        unmatched_jira_issues = set()
+        unmatched_forge_issues = set(forge_issues)
+
+        for jira_issue in jira_issues:
+            full_url = cls.get_full_url_from_jira_issue(jira_issue)
+            match = None
+            for forge_issue in unmatched_forge_issues:
+                if forge_issue.full_url == full_url:
+                    log.debug(
+                        "%s: Matched with forge issue %s", jira_issue.key, forge_issue.full_url
+                    )
+                    match = (jira_issue, forge_issue)
+                    matched_issues.add(match)
+                    break
+            if match:
+                unmatched_forge_issues.remove(forge_issue)
+            else:
+                log.debug("%s: Unmatched with forge issue", jira_issue.key)
+                unmatched_jira_issues.add(jira_issue)
+
+        return matched_issues, unmatched_jira_issues, unmatched_forge_issues
+
+    def close_jira_issues(self, jira_issues: Collection[JiraIssue]) -> None:
+        """Close JIRA issues.
+
+        :param jira_issues: The JIRA issues to be closed
+        """
+        log.info(
+            "Closing %s JIRA issues: %s",
+            len(jira_issues),
+            ", ".join(jira_issue.key for jira_issue in jira_issues),
+        )
+        for jira_issue in jira_issues:
+            log.info(
+                "%s: Transitioning issue from %s to %s",
+                jira_issue.key,
+                jira_issue.fields.status.name,
+                self._jira_statuses.closed,
+            )
+            self._jira.transition_issue(jira_issue, self._jira_statuses.closed)
+
+    def create_or_reopen_jira_issues(self, forge_issues: Collection[Issue]) -> set[MatchedIssue]:
+        """Create or reopen JIRA issues for issues on forges.
+
+        :param forge_issues: The forge issues which should have their
+            corresponding JIRA issues looked up or created.
+        :return: A set of pairs of matched JIRA and forge issues
+        """
+        if not forge_issues:
+            log.info("No JIRA issues to create or reopen.")
+            return set()
+
+        log.info("Creating/reopening JIRA issues for unmatched forge issues…")
+
+        log.info("Retrieving list of closed JIRA issues…")
+        closed_jira_issues = self._jira.get_issues_by_labels(self._jira_config.label, closed=True)
```

Comment:
This could take a long time as there will be more and more closed issues with the `infra-releng` label over time. But I'm not sure how to do this better.

Reply:
I’ve tried it out and right now it’s bearable 😅. We could also filter by `<forge>:<repo>` labels additionally, but it would need more flexibility of the `get_issues_by_labels()` method – right now it only combines labels with AND, we’d need something like `<global label> AND (<forge:repo1> OR <forge:repo2> OR …)`.

Reply:
Let’s do this in a follow-up PR.

Reply:
⇒ #111